packages/vite/src/node/cli.ts

```diff
@@ -175,93 +177,148 @@ cli
     '--force',
     `[boolean] force the optimizer to ignore the cache and re-bundle`,
   )
-  .action(async (root: string, options: ServerOptions & GlobalCLIOptions) => {
-    filterDuplicateOptions(options)
-    // output structure is preserved even after bundling so require()
-    // is ok here
-    const { createServer } = await import('./server')
-    try {
-      const server = await createServer({
-        root,
-        base: options.base,
-        mode: options.mode,
-        configFile: options.config,
-        configLoader: options.configLoader,
-        logLevel: options.logLevel,
-        clearScreen: options.clearScreen,
-        server: cleanGlobalCLIOptions(options),
-        forceOptimizeDeps: options.force,
-      })
-
-      if (!server.httpServer) {
-        throw new Error('HTTP server not available')
+  .option(
+    '--fullBundleMode',
+    `[boolean] Enable bundle at dev to instead of esm dev server`,
+  )
+  // TODO(underfin): Consider how to merge the build option into dev command.
+  .action(
+    async (
+      root: string,
+      options: BuildEnvironmentOptions &
+        BuilderCLIOptions &
+        ServerOptions &
+        GlobalCLIOptions,
+    ) => {
+      filterDuplicateOptions(options)
+      // output structure is preserved even after bundling so require()
+      // is ok here
+      async function createServerWithFullBundleMode() {
+        const { createServerWithResolvedConfig } = await import('./server')
+
+        const { createBuilder } = await import('./build')
+
+        const buildOptions: BuildEnvironmentOptions =
+          cleanBuilderCLIOptions(options)
+
+        const inlineConfig: InlineConfig = {
+          root,
+          base: options.base,
+          mode: options.mode,
+          configFile: options.config,
+          configLoader: options.configLoader,
+          logLevel: options.logLevel,
+          clearScreen: options.clearScreen,
+          build: buildOptions,
+          experimental: {
+            fullBundleMode: true,
+          },
+          ...(options.app ? { builder: {} } : {}),
+        }
+        const builder = await createBuilder(inlineConfig, null, 'serve')
+
+        const server = await createServerWithResolvedConfig(builder.config)
+
+        if (!server.httpServer) {
+          throw new Error('HTTP server not available')
+        }
```

Comment:
We should probably support middleware mode (it does not have `server.httpServer`) in the future.